expression/builtin_miscellaneous.go

```diff
@@ -528,106 +525,102 @@ type isIPv4CompatFunctionClass struct {
 }
 
 func (c *isIPv4CompatFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinIsIPv4PrefixedSig{newBaseBuiltinFunc(args, ctx), false}
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sig := &builtinIsIPv4PrefixedSig{baseIntBuiltinFunc{bf}, false}
+	return sig.setSelf(sig), nil
+}
+
+type isIPv4MappedFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *isIPv4MappedFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sig := &builtinIsIPv4PrefixedSig{baseIntBuiltinFunc{bf}, true}
+	return sig.setSelf(sig), nil
 }
 
 type builtinIsIPv4PrefixedSig struct {
-	baseBuiltinFunc
+	baseIntBuiltinFunc
 	// isIPv4MappedPrefix true for `Is_IPv4_Mapped`, false for `Is_IPv4_Compat`
 	isIPv4MappedPrefix bool
 }
 
-// eval evals Is_IPv4_Mapped or Is_IPv4_Compat
+// evalInt evals Is_IPv4_Mapped or Is_IPv4_Compat
 // See https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_is-ipv4-compat
 // See https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_is-ipv4-mapped
-func (b *builtinIsIPv4PrefixedSig) eval(row []types.Datum) (d types.Datum, err error) {
+func (b *builtinIsIPv4PrefixedSig) evalInt(row []types.Datum) (int64, bool, error) {
+	val, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if err != nil || isNull {
+		return int64(0), false, errors.Trace(err)
+	}
 	var (
 		prefixMapped = []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0xff, 0xff}
 		prefixCompat = []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
 	)
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
-	}
-
-	arg := args[0]
-	d.SetInt64(0)
-	if arg.IsNull() {
-		return d, nil
-	}
-
-	ipAddress, err := arg.ToBytes()
-	if err != nil {
-		return d, errors.Trace(err)
-	}
 
+	ipAddress := []byte(val)
 	if len(ipAddress) != net.IPv6len {
 		//Not an IPv6 address, return false
-		return
+		return int64(0), false, nil
 	}
 
 	if b.isIPv4MappedPrefix {
```

Comment:
there are two ways to avoid this `if...else...` branch:

1. write `builtinIsIPv4MappedPrefixedSig` and `builtinIsIPv4CompatPrefixedSig`, implement `evalInt()` respectively, or:
2. replace `isIPv4MappedPrefix bool` field with a function pointer, implement two functions for mapped and compat respectively, set the function pointer to the proper function in `getFunction()` of `isIPv4MappedFunctionClass` and `isIPv4CompatFunctionClass`